opentech/apply/funds/forms.py

```diff
@@ -165,13 +168,31 @@ def save(self, *args, **kwargs):
             if reviewer not in instance.reviewers.filter(assignedreviewers__role=None)
         )
 
+        # Update or create role reviewers
         for role, reviewer in assigned_roles.items():
             if reviewer:
-                AssignedReviewers.objects.update_or_create(
+                # Edge case: handle any role reviewers that reviewed and now have been removed from role assignment,
+                # We want to have a AssignedReviewer record for that orphaned reviewer
+                # because they have left a review but no longer have a role
+                existing = AssignedReviewers.objects.filter(
```

Comment:
I think this is now too complex for what it needs to do, I think by excluding the roles from `submitted_users` before we create the new objects we are causing a headache. 

What we need to do is:
* Update role reviewers 
* Update non-reviewers 
   * remove people not on form
   * add people from form
* Add in anyone who has already reviewed

So the logic is:

* Users who need non role assignment `non_role` = set([submitted reviewers, assigned staff - no role, external reviewers])
* Update new assigned reviewers (as original) 
* Delete users without a role if they don't appear in `non_role` 
* Create new users without roles from `non_roles` if they do not appear in `assigned(role=None)`

The logic to handle adding users once they have reviewed is part of the post-save method in the `reviews/models.py`